packages/examples/packages/ledger/src/transport.ts

```diff
@@ -0,0 +1,240 @@
+     * @param device - The device to emit.
+     */
+    function emit(device: HidDevice) {
+      observer.next({
+        type: 'add',
+        descriptor: device,
+        deviceModel: identifyUSBProductId(device.productId),
+      });
+
+      observer.complete();
+    }
+
+    this.list()
+      .then((devices) => {
+        if (unsubscribed) {
+          return;
+        }
+
+        if (devices.length > 0) {
+          emit(devices[0]);
+          return;
+        }
+
+        requestDevice()
+          .then((device) => {
+            if (unsubscribed) {
+              return;
+            }
+
+            if (!device) {
+              observer.error(
+                new TransportOpenUserCancelled(
+                  'No device was provided to connect to.',
+                ),
+              );
+
+              return;
+            }
+
+            emit(device);
+          })
+          .catch((error) => {
+            observer.error(new TransportOpenUserCancelled(error.message));
+          });
+      })
+      .catch((error) => {
+        observer.error(new TransportOpenUserCancelled(error.message));
+      });
+
+    return { unsubscribe };
+  }
+
+  /**
+   * Request to connect to a Ledger device. This will always prompt the user to
+   * connect a device.
+   *
+   * @returns A promise that resolves to a transport.
+   */
+  static async request() {
+    const device = await requestDevice();
+    if (!device) {
+      throw new TransportOpenUserCancelled(
+        'No device was provided to connect to.',
+      );
+    }
+
+    return this.open(device);
+  }
+
+  /**
+   * Create a transport with a previously connected device. Returns `null` if no
+   * device was found.
+   *
+   * @returns A promise that resolves to a transport, or `null` if no device was
+   * found.
+   */
+  static async openConnected() {
+    const devices = await this.list();
+    if (devices.length > 0) {
+      return this.open(devices[0]);
+    }
+
+    return null;
+  }
+
+  /**
+   * Create a transport with a specific device.
+   *
+   * @param device - The device to connect to.
+   * @returns A transport.
+   */
+  static async open(device: HidDevice) {
+    return new TransportSnapsHID(device);
+  }
+
+  /**
+   * Close the connection to the transport device.
+   */
+  async close() {
+    // Snaps devices cannot be closed.
+  }
+
+  /**
+   * Exchange with the device using APDU protocol.
+   *
+   * @param apdu - The APDU command to send to the device.
+   * @returns The response from the device.
+   */
+  exchange = async (apdu: Buffer): Promise<Buffer> => {
+    return await this.exchangeAtomicImpl(async () => {
+      const framing = hidFraming(this.#channel, this.#packetSize);
+      const blocks = framing.makeBlocks(apdu);
+
+      for (const block of blocks) {
+        await snap.request({
+          method: 'snap_writeDevice',
+          params: {
+            type: 'hid',
+            id: this.device.id,
+            data: bytesToHex(block),
```

Comment:
How large are these blocks? It might help us make a decision on the type for `data`